net: socket: asynchronous connect

neja20 opened the pull request:
Added a feature of socket connect
being asynchronous. If socket is set
to nonblock with O_NONBLOCK flag,
then connect() is non-blocking aswell.
App can normally poll the socket to
test when the connection is established.

Signed-off-by: Daniel Nejezchleb <dnejezchleb@hwg.cz>

rlubos reviewed (commented):
The changes you propose look reasonable, however, they only seem to cover the positive path (i e. connection is established). It seems to be that some extra processing is needed if the connection fails. Currently, no action is taken wrt `connect_sem`, so the `poll()` call monitoring the socket could block endlessly. Plus, I think that connect callback should also be called (there is a status field) for completeness.

I think it'd be good to add additional tests to `tests/net/sockets/tcp` to verify that the asynchronous connect works correctly.

Finally, we'd need to align TLS implementation as well - given that such a socket does not support non-blocking handshake anyway, we could force using blocking connect as well (by temporary lifting the O_NONBLOCK flag on the underlying socket).

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -492,11 +504,32 @@ int zsock_connect_ctx(struct net_context *ctx, const struct sockaddr *addr,
+		if (sock_is_error(ctx)) {
+			int err = ETIMEDOUT;
+
+			if (ctx->user_data) {
```
Not sure if I understand this check, shouldn't `user_data` contain an error code regardless, if the socket is in error state?

rlubos reviewed (commented):
A little bit more feedback. Generally, it looks like a solution that could work, however, I'd really feel more comfortable about this if we had a test case for a non-blocking connection in a positive and negative scenario.

teburd commented:
@neja20 please take a look at the comments from @rlubos

neja20 commented:
@teburd yes i know about it. I am currently busy with other work on our project. I will get back to it as soon as i am done with that.

neja20 commented on subsys/net/ip/tcp.c:
```diff
@@ -415,7 +415,20 @@ static int tcp_conn_unref(struct tcp *conn, int status)
 		tcp_send_queue_flush(conn);
-		goto out;
+
+		if (conn->async_connect) {
```
I see the problem of feeding  `connect_sem` in a sychronized case when it would wake up the thread waiting for `connect_sem` in `net_tcp_connect` and then it would race in the `tcp_conn_unref` . Though in that case, there may be a solution by moving `tcp_lock` to the top of that function. What do you think?

rlubos replied:
It seems that the code around the `in_connect` flag could use some extra mutex protection anyway. It seems to me that a similar race could take place when you close the socket that connects asynchronously before the connection is established.

neja20 replied:
Ok I used `conn->lock` to protect that section

neja20 commented:
Please take a look at the changes I implemented upon your comments. 
I also lifted the O_NONBLOCK in TLS connect and  added the test although I don't know if it has the form that has been desired. Thank you.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -901,6 +901,104 @@ ZTEST(net_socket_tcp, test_connect_timeout)
+	int rv;
+	struct pollfd poll_fds[POLL_FDS_NUM];
+
+	zassert_equal(wait_for_n_tcp_contexts(0, TCP_TEARDOWN_TIMEOUT),
+		      0,
+		      "Not all TCP contexts properly cleaned up");
```
This shouldn't really be needed before the test, its' a teardown procedure.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -901,6 +901,104 @@ ZTEST(net_socket_tcp, test_connect_timeout)
+	struct pollfd poll_fds[1];
+	int poll_rc;
+
+	zassert_equal(wait_for_n_tcp_contexts(0, TCP_TEARDOWN_TIMEOUT),
+		      0,
+		      "Not all TCP contexts properly cleaned up");
+
+	restore_packet_loss_ratio();
```
Same here, also applies to `restore_packet_loss_ratio();`.

rlubos reviewed (commented):
One final request - please split tests and code changes into separate commits.

I'm still not sure if `async_connect` flag is really needed, as it seems we could differentiate between timeout/abort in `net_tcp_connect()` to avoid calling unref twice on a TCP context. But the test cases make sense and they pass so I guess it's fine.

dnejezchleb committed:
tests: net: socket: asynchronous connect

Added tests to check async socket connect

Signed-off-by: Daniel Nejezchleb <dnejezchleb@hwg.cz>

dnejezchleb committed:
net: socket: asynchronous connect

Added a feature of socket connect
being asynchronous. If socket is set
to nonblock with O_NONBLOCK flag,
then connect() is non-blocking aswell.
App can normally poll the socket to
test when the connection is established.

Signed-off-by: Daniel Nejezchleb <dnejezchleb@hwg.cz>

neja20 commented:
I didn't like `async_connect` either. So I removed it and basically used `in_connect` to protect against second unref in `net_tcp_connect()`
